src/sys/fuchsia/selector.rs

```diff
@@ -0,0 +1,355 @@
+use {io, Event, PollOpt, Ready, Token};
+use sys::fuchsia::{
+    epoll_event_to_ready,
+    poll_opts_to_wait_async,
+    status_to_io_err,
+    EventedFd,
+    EventedFdInner,
+};
+use concurrent_hashmap::ConcHashMap;
+use magenta;
+use magenta_sys;
+use magenta::HandleBase;
+use std::collections::hash_map::RandomState;
+use std::fmt;
+use std::mem;
+use std::sync::atomic::{AtomicBool, AtomicUsize, ATOMIC_USIZE_INIT, Ordering};
+use std::sync::{Arc, Mutex, Weak};
+use std::time::Duration;
+use sys;
+
+/// The kind of registration-- file descriptor or handle.
+///
+/// The last bit of a token is set to indicate the type of the registration.
+#[derive(Copy, Clone, Eq, PartialEq)]
+enum RegType {
+    Fd,
+    Handle,
+}
+
+fn key_from_token_and_type(token: Token, reg_type: RegType) -> u64 {
+    let key = token.0 as u64;
+    let msb = 1u64 << 63;
+    if (key & msb) != 0 {
+        panic!("Most-significant bit of token must remain unset.");
+    }
+
+    match reg_type {
+        RegType::Fd => key,
+        RegType::Handle => key | msb,
+    }
+}
+
+fn token_and_type_from_key(key: u64) -> (Token, RegType) {
+    let msb = 1u64 << 63;
+    (
+        Token((key & !msb) as usize),
+        if (key & msb) == 0 {
+            RegType::Fd
+        } else {
+            RegType::Handle
+        }
+    )
+}
+
+/// Each Selector has a globally unique(ish) ID associated with it. This ID
+/// gets tracked by `TcpStream`, `TcpListener`, etc... when they are first
+/// registered with the `Selector`. If a type that is previously associated with
+/// a `Selector` attempts to register itself with a different `Selector`, the
+/// operation will return with an error. This matches windows behavior.
+static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+
+pub struct Selector {
+    id: usize,
+
+    /// Magenta object on which the handles have been registered, and on which events occur
+    port: Arc<magenta::Port>,
+
+    /// Whether or not `tokens_to_rereg` contains any elements. This is a best-effort attempt
+    /// used to prevent having to lock `tokens_to_rereg` when it is empty.
+    has_tokens_to_rereg: AtomicBool,
+
+    /// List of `Token`s corresponding to registrations that need to be reregistered before the
+    /// next `port::wait`. This is necessary to provide level-triggered behavior for
+    /// `Async::repeating` registrations.
+    ///
+    /// When a level-triggered `Async::repeating` event is seen, its token is added to this list so
+    /// that it will be reregistered before the next `port::wait` call, making `port::wait` return
+    /// immediately if the signal was high during the reregistration.
+    tokens_to_rereg: Mutex<Vec<Token>>,
+
+    /// Map from tokens to weak references to `EventedFdInner`-- a structure describing a
+    /// file handle, its associated `mxio` object, and its current registration.
+    token_to_fd: ConcHashMap<Token, Weak<EventedFdInner>>,
```

Comment:
Did you find this necessary for performance? (over just a `Mutex<HashMap<...>>`) With the liberal usage of `Mutex` elsewhere here I'd be tempted to avoid picking up this dependency and just stick to a simple construction for now.